[SYCL] Add code location data to XPTI trace in case of exception thrown

KseniyaTikhomirova opened the pull request:
Adds extra information about code location of command submission to exception message.
Now instrumented a few sycl::queue non-variadic methods which takes code location from parameter.
Also instrumented variadic parallel_for where code location is extracted from kernel info.

KseniyaTikhomirova commented on sycl/include/sycl/queue.hpp:
```diff
@@ -1439,8 +1445,12 @@ class __SYCL_EXPORT queue : public detail::OwnerLessBase<queue> {
+    constexpr detail::code_location CodeLoc(
+        KI::getFileName(), KI::getFunctionName(), KI::getLineNumber(),
+        KI::getColumnNumber());
+    std::cout << "KI::getFileName() = " << KI::getFileName() << std::endl;
```
remove print

KseniyaTikhomirova commented on sycl/source/CMakeLists.txt:
```diff
@@ -3,7 +3,7 @@
 #cmake_policy(SET CMP0057 NEW)
 #include(AddLLVM)
-
+add_definitions(-gdwarf-4 -O0)
```
remove

KseniyaTikhomirova commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -66,6 +66,44 @@ std::atomic_uint ObjectUsageCounter::MCounter{0};
 
+void GlobalHandler::InitXPTIStuff() {
+  // Let subscribers know a new stream is being initialized
+  getXPTIRegistry().initializeStream(SYCL_SYCLCALL_STREAM_NAME, GMajVer,
```
change the name, it is ugly

KseniyaTikhomirova commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -66,6 +66,44 @@ std::atomic_uint ObjectUsageCounter::MCounter{0};
+                    xpti_at::active, &SYCLInstanceNo);
+}
+
+bool hasValue(const char *Value) { return Value && Value[0] != '\0'; }
```
not used here any more

KseniyaTikhomirova commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -66,6 +66,44 @@ std::atomic_uint ObjectUsageCounter::MCounter{0};
+                            CodeLocation.fileName(), CodeLocation.lineNumber(),
+                            CodeLocation.columnNumber(), nullptr);
+
+    uint64_t EventInstanceNo;
```
initialize

KseniyaTikhomirova commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -66,6 +66,44 @@ std::atomic_uint ObjectUsageCounter::MCounter{0};
 GlobalHandler::~GlobalHandler() = default;
 
+void GlobalHandler::InitXPTIStuff() {
+  // Let subscribers know a new stream is being initialized
```
add define if xpti macro defined

KseniyaTikhomirova commented on sycl/source/detail/scheduler/commands.cpp:
```diff
@@ -296,16 +305,23 @@ class DispatchHostTask {
     assert(MThisCmd->getCG().getType() == CG::CGTYPE::CodeplayHostTask);
 
     CGHostTask &HostTask = static_cast<CGHostTask &>(MThisCmd->getCG());
+#ifdef XPTI_ENABLE_INSTRUMENTATION
```
to add comment for this magic

KseniyaTikhomirova commented on sycl/unittests/xpti_trace/xptitest_subscriber/XPTISubscriber.cpp:
```diff
@@ -0,0 +1,65 @@
+  xptiRegisterCallback(StreamID, xpti::trace_diagnostics, testCallback);
+}
+
+XPTI_CALLBACK_API void xptiTraceFinish(const char *StreamName) {}
```
unregisterSTream? to check if I need to properly shutdown it in global_handler too.

KseniyaTikhomirova commented on xpti/CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
 cmake_minimum_required(VERSION 3.8)
-
+add_definitions(-gdwarf-4 -O0)
```
remove

KseniyaTikhomirova commented on xpti/include/xpti/xpti_trace_framework.h:
```diff
@@ -457,6 +457,8 @@ xptiQueryMetadata(xpti::trace_event_data_t *e);
 /// @return bool that indicates whether it is enabled or not
 XPTI_EXPORT_API bool xptiTraceEnabled();
 
+XPTI_EXPORT_API void xptiTraceTryToEnable();
```
do we need any ABI tests for xpti to be updated with new symbols?

KseniyaTikhomirova replied:
need to align with Vasanth about these changes.

KseniyaTikhomirova commented on xpti/src/xpti_proxy.cpp:
```diff
@@ -40,6 +40,7 @@ enum functions_t {
   XPTI_TRACE_ENABLED,
   XPTI_REGISTER_PAYLOAD,
   XPTI_QUERY_PAYLOAD_BY_UID,
+  // XPTI_TRACE_TRY_TO_ENABLE,
```
remove

KseniyaTikhomirova commented on xpti/src/xpti_proxy.cpp:
```diff
@@ -74,7 +75,8 @@ class ProxyLoader {
       {XPTI_QUERY_METADATA, "xptiQueryMetadata"},
-      {XPTI_TRACE_ENABLED, "xptiTraceEnabled"}};
+      {XPTI_TRACE_ENABLED, "xptiTraceEnabled"},
+      /*{XPTI_TRACE_TRY_TO_ENABLE, "xptiTraceTryToEnable"}*/};
```
remove

KseniyaTikhomirova commented on xpti/src/xpti_proxy.cpp:
```diff
@@ -139,6 +115,38 @@ class ProxyLoader {
     return *loader;
   }
+  // needed for testing
+  void tryToEnable() {
```
need to discuss with Vasanth. Here it is one way method which could check and enable in runtime. Do we need an opportunity to disable it in runtime too?

tovinkere replied:
> need to discuss with Vasanth. Here it is one way method which could check and enable in runtime. Do we need an opportunity to disable it in runtime too?

We seem to be able to enable/disable at runtime already. Currently, setting XPTI_TRACE_ENABLE=0 will disable the instrumentation as xptiTraceEnabled() will fail. Are you referring to something different? Are you referring to being able to do this programmatically?

KseniyaTikhomirova replied:
@tovinkere yes, I need to have an option to enable traces in runtime. Now xpti checks env vars at loading and no way to update status after.

KseniyaTikhomirova commented on xptifw/src/CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
 file(GLOB SOURCES *.cpp)
-
+add_definitions(-gdwarf-4 -O0)
```
remove

KseniyaTikhomirova replied:
done

KseniyaTikhomirova commented on xptifw/src/xpti_trace_framework.cpp:
```diff
@@ -959,9 +959,10 @@ class Framework {
       // provided and the trace_point_type is function_begin/function_end.
       // This allows us to trace function calls without too much effort.
-      std::array<trace_point_type_t, 13> AllowedTypes = {
+      std::array<trace_point_type_t, 14> AllowedTypes = {
```
also to clarify with Vasanth if there was an idea that could block me on there changes.

KseniyaTikhomirova commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1545

KseniyaTikhomirova commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1545

KseniyaTikhomirova commented:
@tovinkere kindly ping

tovinkere commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -66,6 +66,44 @@ std::atomic_uint ObjectUsageCounter::MCounter{0};
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+  if (xptiTraceEnabled()) {
+    uint64_t Uid = xptiGetUniqueId(); // Gets the UID from TLS
+    uint8_t StreamID = xptiRegisterStream(SYCL_SYCLCALL_STREAM_NAME);
```
I think diagnostics from exceptions should always be available for any tool.  Most subscribers will always subscribe to "sycl" stream and it makes sense for this stream to be used. Using "sycl.debug" stream for this maybe okay too. Creating too many new streams will create a problem for subscribers that already exist and are subscribing from the "sycl" stream or streams that existed when they created a subscriber. They have to refactor their subscriber to get this data. By using "sycl" we can avoid that and they will continue to get new data when it's available.

KseniyaTikhomirova replied:
sorry, didn't catch the point. what exactly requires improvement here?

KseniyaTikhomirova replied:
fixed

tovinkere commented on sycl/source/detail/scheduler/commands.hpp:
```diff
@@ -329,6 +337,14 @@ class Command {
   bool MFirstInstance = false;
   /// Instance ID tracked for the command.
   uint64_t MInstanceID = 0;
+  /// Code location of command submission to sycl is available.
+  /// Assigned with the valid value only if command execution is async (host
+  /// task) or delayed (blocked by host task).
+  code_location MSubmissionCodeLocation;
+  /// Introduces string to handle memory management since code_location struct
+  /// works with raw car arrays.
+  std::string MSubmissionFileName;
+  std::string MSubmissionFunctionName;
```
Will this break ABI compatibility?

KseniyaTikhomirova replied:
no, it is internal class, not SYCL API

tovinkere commented on sycl/source/detail/scheduler/commands.cpp:
```diff
@@ -2554,6 +2617,10 @@ pi_int32 ExecCGCommand::enqueueImp() {
 
+    // Host task is executed asynchronously so we should record where it was
+    // submitted to report exception origin properly.
+    copySubmissionCodeLocation();
```
Can this not be done in the base class? The base class constructor is also getting "CodeLoc" from the submit call, so you may be able to consolidate this and call copySubmissionCodeLocation() there.

KseniyaTikhomirova replied:
I did it here because now it is the only type of command that deal with it. And no need to do this operation for all other classes.

tovinkere commented on sycl/source/detail/scheduler/graph_processor.cpp:
```diff
@@ -59,6 +59,10 @@ bool Scheduler::GraphProcessor::handleBlockingCmd(Command *Cmd,
+
+      // Blocked command will be enqueued asynchronously from submission so we
+      // need to keep current root code location to report failure properly.
+      RootCommand->copySubmissionCodeLocation();
```
Are we trying to copy the TLS code location info to RootCommand? Shouldn't it already be there if we had it in the constructor?

KseniyaTikhomirova replied:
we keep it in the constructor for host task only. Although if kernel depends on host task - it will be blocked by it and enqueued in host task thread after host kernel completion. To handle it I copy code location here.

tovinkere commented on sycl/source/detail/xpti_registry.hpp:
```diff
@@ -39,12 +40,24 @@ extern uint8_t GBufferStreamID;
 inline constexpr const char *SYCL_BUFFER_STREAM_NAME =
     "sycl.experimental.buffer";
 
+inline constexpr const char *SYCL_SYCLCALL_STREAM_NAME = "sycl.api";
```
Prefer we do not create new streams - better to categorize them and add them to existing streams to support subscribers that have already been implemented. I see that you are trying to map the exception output stream (sycl.api) to SYCL_TRACE_API_ENABLE.

KseniyaTikhomirova replied:
fixed

tovinkere commented on xpti/src/xpti_proxy.cpp:
```diff
@@ -74,7 +75,8 @@ class ProxyLoader {
       {XPTI_ADD_METADATA, "xptiAddMetadata"},
       {XPTI_QUERY_METADATA, "xptiQueryMetadata"},
-      {XPTI_TRACE_ENABLED, "xptiTraceEnabled"}};
+      {XPTI_TRACE_ENABLED, "xptiTraceEnabled"},
```
Extra comment since the new XPTI API has been commented out? If a new API is added, it must be implemented in the proxy library and the dispatcher.

KseniyaTikhomirova replied:
thanks for the guidance, will update

KseniyaTikhomirova replied:
fixed. added xptiForceSetTraceEnabled. xptiTryToEnable is implemented in proxy only and is not needed to be implemented and exposed in xptifw.

tovinkere commented on xpti/src/xpti_proxy.cpp:
```diff
@@ -420,6 +428,10 @@ XPTI_EXPORT_API bool xptiTraceEnabled() {
 }
 
+XPTI_EXPORT_API void xptiTraceTryToEnable() {
+  xpti::ProxyLoader::instance().tryToEnable();
```
I don't think this changes anything, so xptiTraceTryToEnable() is redundant. Everything is controlled by environment variables.

KseniyaTikhomirova replied:
yes, it is controlled by environment variables but no way to enable xpti traces programmatically - set env vars in runtime.

KseniyaTikhomirova replied:
@tovinkere let me to share more details about runtime enabling I want to add. XPTI_TRACE_ENABLE allows to enable traces only if other variables were set at the program loading time. There is no way to requery the state of these variables . If they are set programmatically from code - no chance to use xpti at all. I want to add this functionality to be able to do unit tests.
  unittest::ScopedEnvVar PathToXPTIFW{"XPTI_FRAMEWORK_DISPATCHER",
                                      "libxptifw.so", [] {}};
  unittest::ScopedEnvVar XPTISubscriber{"XPTI_SUBSCRIBERS",
                                        "libxptitest_subscriber.so", [] {}};

KseniyaTikhomirova replied:
discussed in person and updated impl.

tovinkere commented on xptifw/src/xpti_trace_framework.cpp:
```diff
@@ -959,9 +959,10 @@ class Framework {
+      std::array<trace_point_type_t, 14> AllowedTypes = {
           trace_point_type_t::function_begin,
           trace_point_type_t::function_end,
+          trace_point_type_t::diagnostics,
```
This will break ABI compatibility. Adding it to the end as the last element may keep compatibility.

KseniyaTikhomirova replied:
will update, thanks

KseniyaTikhomirova replied:
fixed

tovinkere commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -67,38 +67,39 @@ GlobalHandler::GlobalHandler() = default;
-                          Event, Uid, static_cast<const void *>(Message));
+    xptiNotifySubscribers(StreamID, xpti::trace_diagnostics,
+                          static_cast<xpti_td *>(GSYCLCallEvent), Event, Uid,
+                          static_cast<const void *>(Message));
```
// We have to handle the cases where: (1) we may have just the code location set 
// and not UID and (2) UID set
detail::tls_code_loc_t Tls;
auto CodeLocation= Tls.query();

// Creating a tracepoint will convert a CodeLocation to UID, if not set
xpti::framework::tracepoint_t TP(CodeLocation.fileName(), CodeLocation.functionName(), CodeLocation.lineNumber(), CodeLocation.ColumnNumber(), nullptr);

// The call to notify will have the signature of:
//   (1) the stream defined in .stream()
//   (2) The trace type equal to what is set by .trace_type()
//   (3) Parent event set to NULL
//   (4) Current event set to one created from CodeLocation and UID
//   (5) An instance ID that records the number of times this code location has been seen
//   (6) The message generated by the exception handler
TP.stream(SYCL_API_STREAM_NAME)
    .trace_type(xpti::trace_diagnostics)
    .event_type(xpti::trace_event_type_t::algorithm)
    .notify(static_cast<const void *>(Message));

KseniyaTikhomirova replied:
fixed

tovinkere commented on sycl/source/detail/xpti_registry.hpp:
```diff
@@ -56,7 +56,7 @@ constexpr const char *GVerStr = SYCL_VERSION_STR;
     "sycl.experimental.buffer";
 
-inline constexpr const char *SYCL_SYCLCALL_STREAM_NAME = "sycl.api";
+inline constexpr const char *SYCL_API_STREAM_NAME = "sycl.api";
```
It is preferred that the exception messages be captured in the same stream as the SYCL stream (SYCLSTREAM_NAME).

tovinkere reviewed (changes requested):
Some minor changes needed to ensure we handle all cases. The XPTI refactor requested will simplify this in the future.

KseniyaTikhomirova commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -67,38 +67,39 @@ GlobalHandler::GlobalHandler() = default;
   // Let subscribers know a new stream is being initialized
-  getXPTIRegistry().initializeStream(SYCL_SYCLCALL_STREAM_NAME, GMajVer,
-                                     GMinVer, GVerStr);
+  getXPTIRegistry().initializeStream(SYCL_API_STREAM_NAME, GMajVer, GMinVer,
```
fixed

KseniyaTikhomirova commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -67,38 +67,39 @@ GlobalHandler::GlobalHandler() = default;
-                                     GMinVer, GVerStr);
+  getXPTIRegistry().initializeStream(SYCL_API_STREAM_NAME, GMajVer, GMinVer,
+                                     GVerStr);
   xpti::payload_t SYCLPayload("SYCL Interface Layer");
```
fixed

KseniyaTikhomirova commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -67,38 +67,39 @@ GlobalHandler::GlobalHandler() = default;
   xpti::payload_t SYCLPayload("SYCL Interface Layer");
   uint64_t SYCLInstanceNo;
   GSYCLCallEvent =
       xptiMakeEvent("SYCL API Layer", &SYCLPayload, xpti::trace_algorithm_event,
```
fixed

KseniyaTikhomirova commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -71,6 +71,49 @@ std::atomic_uint ObjectUsageCounter::MCounter{0};
+    detail::tls_code_loc_t Tls;
+    auto CodeLocation = Tls.query();
+
+    // Creating a tracepoint will convert a CodeLocation to UID, if not set
```
Hi @tovinkere I would suggest to left responsibility how to treat absence of data to subscribers. Adding extra strings to this data brings more contracts between rt and xpti subscribers which I think could lead to confusion and necessity to align them each time we make any change.
If you will take a look at unit tests subscriber implementation - it handles it by its own and decides what to print. The same is for sycl-trace subscriber.

tovinkere replied:
If we are sending in invalid event types on occasion where code location is not available, we should be documenting it or the "sycl" stream as one of the possibilities for the trace type in question so the callback handler can handle it. All other trace types that are a part of this stream will be sending in valid events, if I am not mistaken. Either we fix it in code or we add documentation.

KseniyaTikhomirova replied:
@tovinkere why is it invalid event? If we not specify code location - event should be still valid. We still could trace exception using its message. When subscriber receives event it will check if it has code location since we have flags to check. Now we could not guarantee that any exception sent will contain code location - at least because not all APIs has instrumentation. But it is not the reason to not trace exceptions at all.

tovinkere replied:
@KseniyaTikhomirova The section in the [documentation ](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/SYCLInstrumentationUsingXPTI.md) for the "sycl" stream should be updated to include expected inputs for xpti::trace_diagnostics.

KseniyaTikhomirova replied:
Thanks for pointing me to that, updated https://github.com/intel/llvm/pull/8101/commits/be352323eee8780bc99fbc7bfee562919225578f 😊

KseniyaTikhomirova commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -79,6 +122,13 @@ GlobalHandler *&GlobalHandler::getInstancePtr() {
 GlobalHandler &GlobalHandler::instance() {
   GlobalHandler *RTGlobalObjHandler = GlobalHandler::getInstancePtr();
   assert(RTGlobalObjHandler && "Handler must not be deallocated earlier");
+
```
InitXPTIStuff is called independently on xptiTryToEnable method. It is called when we query global handler in runtime. So it would be called even if xpti not enabled.

tovinkere replied:
Ok, sounds good!

KseniyaTikhomirova commented on xpti/include/xpti/xpti_trace_framework.h:
```diff
@@ -505,4 +511,5 @@ typedef xpti::result_t (*xpti_add_metadata_t)(xpti::trace_event_data_t *,
                                               const char *, xpti::object_id_t);
 typedef xpti::metadata_t *(*xpti_query_metadata_t)(xpti::trace_event_data_t *);
 typedef bool (*xpti_trace_enabled_t)();
+typedef void (*xpti_force_set_trace_enabled_t)(bool);
```
I left the comment in xptiTraceTryToENable declaration that we need this method now only in proxy. XPTI Framework do not implement this and not need it now.

tovinkere commented on sycl/doc/design/SYCLInstrumentationUsingXPTI.md:
```diff
@@ -256,6 +256,7 @@ All trace point types in bold provide semantic information about the graph, node
 | `wait_end`         | <div style="text-align: left"><li>**trace_type**: `xpti::trace_point_type_t::wait_end` that marks the beginning of the wait on an `event`</li> <li> **parent**: `nullptr`</li> <li> **event**: The event ID will reflect the ID of the command group object submission that created this event or a new event based on the combination of the string "queue.wait" and the address of the ...
 | `barrier_begin`       | <div style="text-align: left"><li>**trace_type**: `xpti::trace_point_type_t::barrier_begin` that marks the beginning of a barrier while enqueuing a command group object</li> <li> **parent**: The global graph event that is created during the `graph_create` event.</li> <li> **event**: The event ID will reflect the ID of the command group object that has encountered a barri ...
 | `barrier_end`         | <div style="text-align: left"><li>**trace_type**: `xpti::trace_point_type_t::barrier_end` that marks the end of the barrier that is encountered during enqueue.</li> <li> **parent**: The global graph event that is created during the `graph_create` event.</li> <li> **event**: The event ID will reflect the ID of the command group object that has encountered a barrier during ...
+| `diagnostics`         | <div style="text-align: left"><li>**trace_type**: `xpti::trace_point_type_t::diagnostics` that represents general purpose notifications. As example, it is emitted when exception is thrown in SYCL runtime. </li> <li> **parent**: Set to NULL.</li> <li> **event**: The event ID will reflect the code location of notification origin if available. </li> <li> **instance**: An in ...
```
`xpti::trace_point_type_t::diagnostics` that represents general purpose notifications. For example, it is emitted when an exception is thrown in SYCL runtime. </li> <li> **parent**: Set to NULL.</li> <li> **event**: The event ID will reflect the code location of notification origin, if available. </li> <li> **instance**: An instance ID that records the number of times this code location has been seen. </li> <li> **user_data**: String with diagnostic message as a `const char *` </li></div> | `sym_function_name`, `sym_source_file_name`, `sym_line_no`, `sym_column_no` |

tovinkere replied:
Made minor edits. Should be good to go.

KseniyaTikhomirova replied:
Hi @tovinkere,
"xpti::trace_point_type_t::diagnostics that represents general purpose notifications. For example, it is emitted when an exception is thrown in SYCL runtime." **with bullet is a common style for all other descriptions. I will break the style unity if do like this.**
FOr other corrections - thanks, I updated that.

KseniyaTikhomirova replied:
https://github.com/intel/llvm/pull/8101/commits/3c38840eda250daeba6ff0fc8212a3285d476ba5

tovinkere reviewed (approved):
LGTM

KseniyaTikhomirova commented:
@llvm-reviewers-runtime, @dpcpp-specification-reviewers, kindly reminder, could you please review?

cperkinsintel commented:
@KseniyaTikhomirova  - can you run this against Jenkins llvm-test-suite ? That's the only way to test it on Windows right now.

KseniyaTikhomirova commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1545

KseniyaTikhomirova commented:
> @KseniyaTikhomirova - can you run this against Jenkins llvm-test-suite ? That's the only way to test it on Windows right now.

I did it on the early stage for this PR (I have a new E2E test ready). Ran it again now. THanks

cperkinsintel commented:
I don't see the Jenkins run in the checks listing here. Am I blind?  I started a run over on https://github.com/intel/llvm-test-suite/pull/1545  and it seems to be running.

KseniyaTikhomirova commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1545

KseniyaTikhomirova commented:
/verify with https://github.com/intel/llvm-test-suite/pull/1545

KseniyaTikhomirova commented:
@cperkinsintel , @intel/dpcpp-specification-reviewers all checks are passed including llvm-test-suite on windows. Could we get it merged, please?

gmlueck commented:
> https://github.com/orgs/intel/teams/dpcpp-specification-reviewers all checks are passed including llvm-test-suite on windows. Could we get it merged, please?

There aren't any changes to the language specs in this PR, and I don't think any of the @intel/dpcpp-specification-reviewers are particularly knowledgeable about the XPTI design.  If you need a review of your changes to the XPTI design doc, it would be better to get a reviewer who is familiar with that code.

KseniyaTikhomirova commented:
> > https://github.com/orgs/intel/teams/dpcpp-specification-reviewers all checks are passed including llvm-test-suite on windows. Could we get it merged, please?
> 
> There aren't any changes to the language specs in this PR, and I don't think any of the @intel/dpcpp-specification-reviewers are particularly knowledgeable about the XPTI design. If you need a review of your changes to the XPTI design doc, it would be better to get a reviewer who is familiar with that code.

Hi, I got approval from @tovinkere who is familiar. I assume I still need formal approval from that reviewers group to merge. Although I am not sure.

steffenlarsen commented:
The dpcpp-specification-reviewers are code-owners of sycl/doc/design/ so in principle their approval is required, otherwise merging requires the gatekeepers to bypass branch protection which, to my knowledge, only a subset of the gatekeepers can do. That said, it is still possible, so for a case like this I would be okay with either someone from the group specifically abstaining (optimally with some reasoning) or by approving it with a note specifying the person or group who's approval they trust in the matter.

If we expect this to be a more frequent problem, we can split the docs into SYCL specification related implementation design (owned by intel/dpcpp-specification-reviewers) and other implementation design documentation (owned by intel/dpcpp-doc-reviewers or intel/llvm-reviewers-runtime).

+@bader for his opinion on both matters.

bader commented:
> The dpcpp-specification-reviewers are code-owners of sycl/doc/design/

It will be better if development team owns the design documentation and dpcpp-specification-reviewers team owns specification extension documents.
dpcpp-specification-reviewers can be manually added to design review if specification is unclear and development team can be asked for feedback on a specification, but it should optional.

tovinkere commented:
> 

Agreed - if not, it adds too many dependencies while merging. Since that is not the case currently, how do we proceed? @gmlueck  Can quickly scan the sycl/doc/design/ changes?

gmlueck reviewed (approved):
There are no language spec changes in this PR, so I have not reviewed anything.  Giving rubber stamp from @intel/dpcpp-specification-reviewers only to allow this PR to proceed.

bader commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -71,6 +71,49 @@ std::atomic_uint ObjectUsageCounter::MCounter{0};
+}
+
+void GlobalHandler::TraceEventXPTI(const char *Message) {
+  if (!Message)
+    return;
+#ifdef XPTI_ENABLE_INSTRUMENTATION
```
```suggestion
#ifdef XPTI_ENABLE_INSTRUMENTATION
  if (!Message)
    return;
```

This code has no value if `XPTI_ENABLE_INSTRUMENTATION` is not set.

KseniyaTikhomirova replied:
done

bader commented on sycl/source/detail/scheduler/commands.cpp:
```diff
@@ -296,16 +305,23 @@ class DispatchHostTask {
-
       // reset host-task's lambda and quit
       HostTask.MHostTask.reset();
+      Scheduler::getInstance().NotifyHostTaskCompletion(MThisCmd);
```
@KseniyaTikhomirova, please, resolve.

KseniyaTikhomirova replied:
created, resolved, thanks

bader commented on sycl/source/detail/scheduler/commands.hpp:
```diff
@@ -303,7 +311,7 @@ class Command {
   std::atomic<EnqueueResultT::ResultT> MEnqueueStatus;
 
-  // All member variable defined here  are needed for the SYCL instrumentation
+  // All member variable defined here are needed for the SYCL instrumentation
```
```suggestion
  // All member variables defined here are needed for the SYCL instrumentation
```

KseniyaTikhomirova replied:
done

bader commented on sycl/source/detail/scheduler/commands.hpp:
```diff
@@ -329,6 +337,14 @@ class Command {
   bool MFirstInstance = false;
   /// Instance ID tracked for the command.
   uint64_t MInstanceID = 0;
+  /// Code location of command submission to sycl is available.
```
What does "to sycl" mean? "to SYCL queue" or "to SYCL API"?

KseniyaTikhomirova replied:
done

bader commented on sycl/source/detail/scheduler/commands.hpp:
```diff
@@ -329,6 +337,14 @@ class Command {
+  /// task) or delayed (blocked by host task).
+  code_location MSubmissionCodeLocation;
+  /// Introduces string to handle memory management since code_location struct
+  /// works with raw car arrays.
```
```suggestion
  /// works with raw char arrays.
```
?

KseniyaTikhomirova replied:
done

bader commented on sycl/tools/sycl-trace/main.cpp:
```diff
@@ -94,6 +101,9 @@ int main(int argc, char **argv, char *env[]) {
     EnablePITrace();
     EnableZETrace();
     EnableCUTrace();
+    // Intentionally do not enable sycl api traces -> to not break existing
```
```suggestion
    // Intentionally do not enable SYCL API traces -> to not break existing
```

bader commented on sycl/tools/sycl-trace/sycl_trace_collector.cpp:
```diff
@@ -0,0 +1,83 @@
+//==----------- sycl_trace_collector.cpp
+//-------------------------------------==//
```
```suggestion
//==----------- sycl_trace_collector.cpp -----------------------------------==//
```

KseniyaTikhomirova replied:
done

bader commented on sycl/tools/sycl-trace/sycl_trace_collector.cpp:
```diff
@@ -0,0 +1,83 @@
+}
+
+// For unification purpose
+void syclPrintersFinish() {}
```
```suggestion
void syclPrintersFinish() {}

```

KseniyaTikhomirova replied:
done

bader commented on sycl/unittests/xpti_trace/xptitest_subscriber/XPTISubscriber.cpp:
```diff
@@ -0,0 +1,66 @@
+//==---------------------- XPTISubscriber.cpp --------------------------==//
```
```suggestion
//==-------------------------- XPTISubscriber.cpp --------------------------==//
```

KseniyaTikhomirova replied:
done

bader commented on sycl/unittests/xpti_trace/xptitest_subscriber/XPTISubscriber.cpp:
```diff
@@ -0,0 +1,66 @@
+                                    xpti::trace_event_data_t *Event,
+                                    uint64_t /*Instance*/,
+                                    const void *UserData) {
+  if (TraceType == xpti::trace_diagnostics) {
```
```suggestion
  if (TraceType != xpti::trace_diagnostics)
    return;

```

Early exist will make it more readable.

KseniyaTikhomirova replied:
done

bader commented on sycl/unittests/xpti_trace/QueueApiFailures.cpp:
```diff
@@ -0,0 +1,502 @@
+//==------------ QueueApiFailures.cpp --- XPTI integration unit tests------==//
```
```suggestion
//==------------ QueueApiFailures.cpp --- XPTI integration unit tests ------==//
```

KseniyaTikhomirova replied:
done

tovinkere reviewed (approved):
LGTM

steffenlarsen commented on sycl/unittests/xpti_trace/QueueApiFailures.cpp:
```diff
@@ -0,0 +1,502 @@
+    ExceptionCaught = true;
+  }
+  Q.wait();
+  EXPECT_TRUE(ExceptionCaught);
```
Should this be an `ASSERT_TRUE` instead? I figure there won't be much point in doing the following checks if the expected error wasn't caught.

KseniyaTikhomirova replied:
thanks, done

steffenlarsen commented on sycl/source/detail/global_handler.cpp:
```diff
@@ -71,6 +71,49 @@ std::atomic_uint ObjectUsageCounter::MCounter{0};
 GlobalHandler::GlobalHandler() = default;
 GlobalHandler::~GlobalHandler() = default;
 
+void GlobalHandler::InitXPTIStuff() {
```
Nit; Although I love the name, I think it would better to omit the `Stuff` part. In my opinion it doesn't add anything to the readability, so better keep it short.

KseniyaTikhomirova replied:
done

steffenlarsen reviewed (approved):
Overall the runtime changes look good! Couple of small comments.

Could you please add a summary to the PR description?

steffenlarsen reviewed (approved):
LGTM!

bader commented:
@KseniyaTikhomirova, please, address this comment from @steffenlarsen:

> Could you please add a summary to the PR description?

I think you missed one suggestion from me - https://github.com/intel/llvm/pull/8101#discussion_r1125029132. Do you mind making this change?

KseniyaTikhomirova commented:
> @KseniyaTikhomirova, please, address this comment from @steffenlarsen:
> 
> > Could you please add a summary to the PR description?
> 
> I think you missed one suggestion from me - [#8101 (comment)](https://github.com/intel/llvm/pull/8101#discussion_r1125029132). Do you mind making this change?

Hi, added description and fixed your comment. Thank you.

KseniyaTikhomirova commented:
> @KseniyaTikhomirova, there are failures in post-commit. Please, address them ASAP. NOTE: there are a few different issues, please, read all logs. https://github.com/intel/llvm/actions/runs/4376461454

working on it, thanks

KseniyaTikhomirova commented:
https://github.com/intel/llvm/pull/8598 waiting for CI checks

bader commented:
> #8598 waiting for CI checks

Thanks. I updated PR title. Please, let me know if you have any comments.

KseniyaTikhomirova commented:
> > #8598 waiting for CI checks
> 
> Thanks. I updated PR title. Please, let me know if you have any comments.

All is perfect. Sorry and thanks a lot.
